docs: synchronize pgmq-rs README with lib.rs doc comments (#535)
tembo-io/pgmq: pgmq-rs/README.md

[![Latest Version](https://img.shields.io/crates/v/pgmq.svg)](https://crates.io/crates/pgmq)

A lightweight message queue. Like [AWS SQS](https://aws.amazon.com/sqs/) and [RSMQ](https://github.com/smrchy/rsmq) but on Postgres.

> The Rust client for PGMQ. This gives you an ORM-like experience with the Postgres extension and makes managing connection pools, transactions, and serialization/deserialization much easier.

Reviewer: These two lines are both trying to serve as the intro to the readme. We should combine/rephrase them so they introduce the PGMQ Postgres extension as a concept first, then introduce this crate as the official Rust client for the extension.

> **Documentation**: https://pgmq.github.io/pgmq/

Reviewer: We should label this as the extension documentation specifically.

> **Source**: https://github.com/pgmq/pgmq

Reviewer: The source is in the `Cargo.toml` and is linked from `crates.io`, so IMO we don't need it in the readme.

## Features

- Lightweight - No background worker or external dependencies, just Postgres SQL objects
- Guaranteed "exactly once" delivery of messages to a consumer within a visibility timeout
- API parity with [AWS SQS](https://aws.amazon.com/sqs/) and [RSMQ](https://github.com/smrchy/rsmq)
- [FIFO](https://github.com/pgmq/pgmq/blob/main/docs/fifo-queues.md#overview) (First-In-First-Out) queues with message group keys for ordered processing
- [Topic-based](https://github.com/pgmq/pgmq/blob/main/docs/topics.md#topic-based-routing) routing with wildcard patterns for publish-subscribe and content-based routing
- Messages stay in the queue until explicitly removed
- Messages can be archived, instead of deleted, for long-term retention and replayability

> - Completely asynchronous API

Reviewer: IMO we don't need the `completely` here.

Supported on Postgres 14-18.

## Table of Contents

- [Postgres Message Queue (PGMQ)](#postgres-message-queue-pgmq)
  - [Features](#features)
  - [Installation](#installation)
    - [Docker](#docker)
    - [SQL Only](#sql-only)
    - [Installing PGMQ via the Rust Client](#installing-pgmq-via-the-rust-client)
      - [Create the DB](#create-the-db)
      - [Initialize applied migrations table](#initialize-applied-migrations-table)
      - [Install using the embedded scripts](#install-using-the-embedded-scripts)
      - [Install using the scripts fetched from GitHub](#install-using-the-scripts-fetched-from-github)
  - [Client Libraries](#client-libraries)
  - [Quick Start](#quick-start)
    - [Minimal example at a glance](#minimal-example-at-a-glance)
  - [Sending messages](#sending-messages)
  - [Reading messages](#reading-messages)
  - [Archive or Delete a message](#archive-or-delete-a-message)
  - [Serialization and Deserialization](#serialization-and-deserialization)
  - [License](#license)

## Installation

> ### Docker

Reviewer: We should include this line from the extension readme:

```
PGMQ can be run on any existing Postgres instance or installed as a Postgres Extension. See [INSTALLATION.md](https://github.com/pgmq/pgmq/blob/main/INSTALLATION.md) for the full installation guide including a [comparison](https://github.com/pgmq/pgmq/blob/main/INSTALLATION.md#considerations) of the Postgres Extension vs the SQL-only installation.
```

The fastest way to get started is by running the Docker image, where PGMQ comes pre-installed as an extension in Postgres.

```bash
docker run -d --name pgmq-postgres -e POSTGRES_PASSWORD=*** -p 5432:5432 ghcr.io/pgmq/pg18-pgmq:v1.10.0
```

Then connect and enable PGMQ:

```bash
psql postgres://postgres:***@localhost:5432/postgres
```

```sql
CREATE EXTENSION pgmq;
```

> ### SQL Only

Reviewer: We should split this up into versioned/unversioned sections like in the main installation instructions.

> 

Reviewer: We should include this warning from the main installation instructions:

```
> ⚠️ This installation approach is not versioned and only works for a fresh installation of `pgmq`.
```

}
```

## Client Libraries

- [Rust](https://github.com/pgmq/pgmq/tree/main/pgmq-rs) (this crate)
- [Python (only for psycopg3)](https://github.com/pgmq/pgmq-py)

Community

- [.NET](https://github.com/brianpursley/Npgmq)
- [Dart](https://github.com/Ofceab-Studio/dart_pgmq)
- [Elixir + Broadway](https://github.com/v0idpwn/off_broadway_pgmq)
- [Elixir](https://github.com/v0idpwn/pgmq-elixir)
- [Go](https://github.com/craigpastro/pgmq-go)
- [Haskell](https://github.com/MichelBoucey/stakhanov)
- [Java (JDBC)](https://github.com/roy20021/pgmq-jdbc-client)

> - [Java (Spring Boot)](https://github.com/adamalexandru4/pgmq-spring)

Reviewer: Any reason you didn't include all of the client libraries from the extension readme?

Also, IMO, this section isn't super necessary.

> ## Quick Start

Reviewer: Please rename this section back to `Examples`

cargo run --example install --features install-sql-github,install-sql-embedded
```

First, you will need Postgres. We use a container in this example.

```bash
docker run -d --name postgres -e POSTGRES_PASSWORD=*** -p 5432:5432 postgres
```

If you don't have Docker installed, it can be found [here](https://docs.docker.com/get-docker/).

Make sure you have the Rust toolchain installed:

```bash
cargo --version
```

This example was written with version 1.67.0, but the latest stable should work. You can go [here](https://www.rust-lang.org/tools/install) to install Rust if you don't have it already, then run `rustup install stable` to install the latest, stable toolchain.

Change directory to the example project:

```bash
cd examples/basic
```

Run the project!

```bash
cargo run
```

> ### Minimal example at a glance

Reviewer 2: I think the minimal example is too long for the readme. Instead just link to the working [examples](https://github.com/pgmq/pgmq/blob/main/pgmq-rs/examples/basic.rs).

Author: Fixed! Replaced the inline minimal example with a link to the working [basic.rs](https://github.com/pgmq/pgmq/blob/main/pgmq-rs/examples/basic.rs) example. Please let me know if anything else is needed.

cargo run --example install --features install-sql-github,install-sql-embedded
```

First, you will need Postgres. We use a container in this example.

```bash
docker run -d --name postgres -e POSTGRES_PASSWORD=*** -p 5432:5432 postgres
```

If you don't have Docker installed, it can be found [here](https://docs.docker.com/get-docker/).

Make sure you have the Rust toolchain installed:

```bash
cargo --version
```

This example was written with version 1.67.0, but the latest stable should work. You can go [here](https://www.rust-lang.org/tools/install) to install Rust if you don't have it already, then run `rustup install stable` to install the latest, stable toolchain.

Change directory to the example project:

```bash
cd examples/basic
```

Run the project!

```bash
cargo run

> ```

Reviewer: IMO this isn't needed -- it's duplicative of the installation section, and very basic Rust stuff

### Minimal example at a glance

See the [basic example](https://github.com/pgmq/pgmq/blob/main/pgmq-rs/examples/basic.rs) for a complete, working example.

## Sending messages

You can send one message at a time with `queue.send()` or several with `queue.send_batch()`.
These methods can be passed any type that implements `serde::Serialize`. This means you can prepare your messages as JSON or as a struct.

## Reading messages

Reading a message will make it invisible (unavailable for consumption) for the duration of the visibility timeout (vt).
No messages are returned when the queue is empty or all messages are invisible.

Messages can be parsed as `serde_json::Value` or into a struct of your design. `queue.read()` returns an `Result<Option<Message<T>>, PgmqError>`

> where `T` is the type of the message on the queue. It returns an error when there is an issue parsing the message (`PgmqError::JsonParsingError`) or if PGMQ is unable to reach postgres (`PgmqError::DatabaseError`).

Reviewer: The text I removed is partially redundant with the note on the next line (and we don't need to specify the DatabaseError part in the readme).

```suggestion
Messages can be parsed as `serde_json::Value` or into a struct of your design. `queue.read()` returns an `Result<Option<Message<T>>, PgmqError>`
where `T` is the type of the message on the queue.
```

Note that when parsing into a `struct`, the operation will return an error if
parsed as the type specified. For example, if the message expected is

> `MyMessage{foo: "bar"}` but `{"hello": "world"}` is received, the application will panic.

Reviewer: ```suggestion
Note that when parsing into a `struct`, the operation will return an error if
the message can not be parsed as the type specified. For example, if the message expected is
`MyMessage{foo: "bar"}` but `{"hello": "world"}` is received, the operation will return an error.
```

Read a single message with `queue.read()` or as many as you want with `queue.read_batch()`.

## Archive or Delete a message

Remove the message from the queue when you are done with it. You can either completely `.delete()`, or `.archive()` the message. Archived messages are deleted from the queue and inserted to the queue's archive table. Deleted messages are just deleted.

> Read messages from the queue archive with SQL:

Reviewer: This reads better (from the extension readme):

```
Archive tables can be inspected directly with SQL. Archive tables have the prefix a_ in the pgmq schema.
```

```sql
SELECT *

> FROM pgmq_{your_queue_name}_archive;

Reviewer: This is not the format of the archive table name. I should be `a_{your_queue_name}`

> ```

Reviewer: It would be nice to have snippets of actual rust code here.
